src/main/java/www/purple/mixxy/controllers/LoginLogoutController.java

```diff
@@ -82,13 +58,69 @@ public Result logout(Context context) {
 
     }
 
-    public Result signup() {
-      return Results.TODO();
+	///////////////////////////////////////////////////////////////////////////
+	// Login
+	///////////////////////////////////////////////////////////////////////////
+    public Result login() {
+    	GoogleAuthHelper helper = new GoogleAuthHelper();
+    	return Results.redirect(helper.buildLoginUrl());
     }
     
-    public Result signupPost() {
-      return Results.TODO();
+    public Result validate(
+    		@Param("state") String state,
+    		@Param("code") String code,
+    		Context context) {
+    	
+    	GoogleAuthHelper helper = new GoogleAuthHelper();
+    	String data = "empty";
+    	String provider = state.substring(0, state.indexOf(';'));
+    	
+    	try {
+			data = helper.getUserInfoJson(code);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+    	
+    	ObjectMapper mapper = new ObjectMapper();
+		try {
+
+			// Convert JSON string to Object
+			GoogleAuthResponse userdata = mapper.readValue(data, GoogleAuthResponse.class);
+			
+			// At this point we can login or signup user
+			String email = userdata.getEmail();
+			String username = email.substring(0, email.indexOf('@'));
+			boolean areCredentialsValid = userDao.isUserValid(username);
+			
+			if(areCredentialsValid) {
+				context.getSession().put("username", userdata.getEmail());
+				context.getFlashScope().success("login.loginSuccessful");
+			} else {
+				// Create new user
+				ObjectifyProvider objectifyProvider = new ObjectifyProvider();
+		        Objectify ofy = objectifyProvider.get();
+		        
+		        // Create a new user and save it
+		        User user = new User(userdata.getEmail(), userdata.getGiven_name(), userdata.getFamily_name(), userdata.getEmail(),
+		        		userdata.getPicture(), userdata.getLocale(), provider, userdata.getId());
+		        ofy.save().entity(user).now();
+		        
+		        context.getSession().put("username", userdata.getEmail());
+				context.getFlashScope().success("login.loginSuccessful");
+				
+		        // Redirect to profile
+				return Results.redirect("/privacy");
+			}
+
+		} catch (JsonGenerationException e) {
```

Comment:
Can you feed these to a `Logger`?  Look at how `UrlNormalizingFilter` does it.  Also, what could go wrong with these exceptions?